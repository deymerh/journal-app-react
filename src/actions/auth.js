import { firebase, gogleAuthProvider } from '../firebase/firebase-config';
import { types } from '../types/types';

export const loginWithEmailAndPassword = (email, password) => {
  return (dispatch) => {
    setTimeout(() => {
      dispatch(login(12345, 'Pedro'))
    }, 3500);
  };
};
export const startGoogleLogin = () => {
  return (dispath) => {
    firebase.auth().signInWithPopup(gogleAuthProvider)
      .then(({ user }) => {
        dispath(login(user.uid, user.displayName))
      })
  }
}
export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName
  }
});