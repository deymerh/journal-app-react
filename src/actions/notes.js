import Swal from 'sweetalert2';
import { db } from '../firebase/firebase-config';
import { fileUpload } from '../helpers/fileUpload';
import { loadNotes } from '../helpers/journalAdd';
import { types } from '../types/types';

export const startNewNote = ()=>{
  return async (dispatch, getState) => {
    const { uid } = getState().auth;
    const newNote = {
      title: '',
      body: '',
      date: new Date().getTime()
    }
    const doc = await db.collection(`${uid}/journal/notes`).add(newNote);
    dispatch(activeNote(doc.id, newNote));
    dispatch(addNewNote(doc.id, newNote));
  }
};

export const startLoadingNotes = (uid)=>{
  return async (dispatch)=>{
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  }
}

export const activeNote = (id, note)=>({
  type: types.notesActive,
  payload: {
    id,
    ...note
  }
});

export const addNewNote = (id, note)=>({
  type: types.notesAddNew,
  payload: {
    id,
    ...note
  }
});

export const setNotes = (notes)=>({
  type:types.notesLoad,
  payload: notes
});

export const startSaveNote = (note)=>{
  return async(dispatch, getState)=>{
    const {uid} = getState().auth;
    if (!note.url) {
      delete note.url;
    }
    const noteToFirestore = {...note};
    delete noteToFirestore.id;
    await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore);
    dispatch(refreshNote(note));
    Swal.fire('Saved', note.title, 'success');
  }
};

export const refreshNote = (note)=>({
  type: types.notesUpdated,
  payload:note
});

export const startUploadImg = (image)=>{
  return async(dispatch, getState)=>{
    const {active:noteActive} = getState().notes;
    Swal.fire({
      title: 'Uploading...',
      text: 'Please wait...',
      allowOutsideClick: false,
      onBeforeOpen: ()=>{
        Swal.showLoading();
      }
    });
    const { secure_url } = await fileUpload(image);
    noteActive.url = secure_url;
    dispatch(startSaveNote(noteActive));
    Swal.close();
  }
};

export const startDeleteNoteSelected = (id)=>{
  return async(dispatch, getState)=>{
    const {uid} = getState().auth;
    await db.doc(`${uid}/journal/notes/${id}`).delete();
    dispatch(deleteNote(id));
  };
};

export const deleteNote = (id)=>({
  type: types.noteSelectDelete,
  payload: id
});