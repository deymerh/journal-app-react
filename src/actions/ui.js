import { types } from "../types/types"

export const setErrorAction = (err) => ({
  type: types.uiSetError,
  payload: err
});

export const setRemoveErrorAction = () => ({
  type: types.uiRemoveError,
});