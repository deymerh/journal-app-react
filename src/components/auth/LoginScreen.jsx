import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import { loginWithEmailAndPassword, startGoogleLogin } from '../../actions/auth';
import { useForm } from '../../hooks/useForm';

export const LoginScreen = () => {
  const dispatch = useDispatch();
  const { loading } = useSelector(state => state.ui);
  const [formsValues, handleInputChange] = useForm({
    email: '',
    password: ''
  });
  const { email, password } = formsValues;
  const handleLogin = (e) => {
    e.preventDefault();
    dispatch(loginWithEmailAndPassword(email, password));
  };
  const handleGoogleLogin = () => {
    dispatch(startGoogleLogin());
  }
  return (
    <div>
      <h1 className="auth__title">Login</h1>
      <form
        onSubmit={handleLogin}
        className="animate__animated animate__fadeIn animated__fater"
      >
        <input
          autoComplete="off"
          className="auth__input"
          name="email"
          onChange={handleInputChange}
          placeholder="Email"
          type="text"
          value={email}
        />
        <input
          autoComplete="off"
          className="auth__input"
          name="password"
          onChange={handleInputChange}
          placeholder="Password"
          type="password"
          value={password}
        />
        <button
          type="submit"
          className="btn btn-primary btn-block"
          disabled={loading}
          data-testid="btnLogin"
        >
          Login
        </button>
        <hr />
        <div className="auth__social-netwoeks">
          <p>Login with social network</p>
          <div
            onClick={handleGoogleLogin}
            className="google-btn"
            data-testid="withGoogle"
          >
            <div className="google-icon-wrapper">
              <img
                className="google-icon"
                src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                alt="google button"
              />
            </div>
            <p className="btn-text">
              <b>Sign in with google</b>
            </p>
          </div>
        </div>
        <Link
          className="link"
          to="/auth/register"
        >
          Create new account
        </Link>
      </form>
    </div>
  );
};
