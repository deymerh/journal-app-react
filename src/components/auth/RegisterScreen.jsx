import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import validator from 'validator';
import { setErrorAction, setRemoveErrorAction } from '../../actions/ui';
import { useForm } from '../../hooks/useForm';
import { registerWithEmailPasswordName } from './../../actions/auth';

export const RegisterScreen = () => {
  const dispatch = useDispatch();
  const { msgError } = useSelector(state => state.ui);
  const [formsValues, handleInputChange] = useForm({
    name: 'Deymer',
    email: 'deymerh@hotmail.com',
    password: "123456",
    password2: "123456",
  });
  const {
    name,
    email,
    password,
    password2
  } = formsValues;
  const handleRegister = (e) => {
    e.preventDefault();
    if (isForm()) {
      dispatch(registerWithEmailPasswordName(email, password, name))
    }
  };
  const isForm = () => {
    if (name.trim().length === 0) {
      dispatch(setErrorAction('Name is required'));
      return false;
    } else if (!validator.isEmail(email)) {
      dispatch(setErrorAction('Email is required'));
      return false;
    } else if (password !== password2 || password.length < 5) {
      dispatch(setErrorAction('Password should be at least 6 characters and match each other'));
      return false;
    }
    dispatch(setRemoveErrorAction());
    return true;
  };
  return (
    <div>
      <h1 className="auth__title">Register</h1>
      <form onSubmit={handleRegister}>
        {
          msgError &&
          (
            <div className="auth__alert-error">{msgError}</div>
          )
        }
        <input
          autoComplete="off"
          className="auth__input"
          name="name"
          onChange={handleInputChange}
          placeholder="Nombre"
          type="text"
          value={name}
        />
        <input
          autoComplete="off"
          className="auth__input"
          name="email"
          onChange={handleInputChange}
          placeholder="Email"
          type="email"
          value={email}
        />
        <input
          autoComplete="off"
          className="auth__input"
          name="password"
          onChange={handleInputChange}
          placeholder="Password"
          type="password"
          value={password}
        />
        <input
          autoComplete="off"
          className="auth__input"
          name="password2"
          onChange={handleInputChange}
          placeholder="Confirm"
          type="password"
          value={password2}
        />
        <button
          className="btn btn-primary btn-block mb-5"
          type="submit"
        >
          Login
        </button>
        <Link
          className="link"
          to="/auth/login"
        >
          Already registered?
        </Link>
      </form>
    </div>
  );
};
