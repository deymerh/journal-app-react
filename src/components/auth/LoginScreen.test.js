import '@testing-library/jest-dom';
import React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { render, fireEvent,  } from '@testing-library/react';
import { LoginScreen } from './LoginScreen';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { startGoogleLogin, loginWithEmailAndPassword } from '../../actions/auth';

jest.mock('../../actions/auth', (()=>({
  startGoogleLogin: jest.fn(),
  loginWithEmailAndPassword:jest.fn(),
})));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {
  auth:{},
  ui:{
    loading: false,
    msgError: null
  }
}
let store = mockStore(initState); 

describe('Pruebas en el <LoginScreen/>', () => {
  beforeEach(()=>{
    store = mockStore(initState);
    jest.clearAllMocks();
  });
  
  test('Debe hacer el Snapshot del componente', () => {
    const component = render(
      <Provider store={store}>
        <MemoryRouter>
          <LoginScreen/>
        </MemoryRouter>
      </Provider>
      );
    expect(component).toMatchSnapshot();
    
  });
  
  test('Clickear al boton de login con google y se ejecute la acción', async() => {
    store.dispatch = jest.fn();
    const {findByTestId} = render(
      <Provider store={store}>
        <MemoryRouter>
          <LoginScreen/>
        </MemoryRouter>
      </Provider>
      );
    const loginGoogle = await findByTestId('withGoogle');
    fireEvent.click(loginGoogle);
    expect(startGoogleLogin).toHaveBeenCalled();
  });

  test('Click login y se ejecute la acción respectiva', async() => {
    store.dispatch = jest.fn();
    const {findByTestId} = render(
      <Provider store={store}>
        <MemoryRouter>
          <LoginScreen/>
        </MemoryRouter>
      </Provider>
      );
    const login = await findByTestId('btnLogin');
    fireEvent.click(login);
    expect(loginWithEmailAndPassword).toHaveBeenCalledWith('','');
  });
  
});
