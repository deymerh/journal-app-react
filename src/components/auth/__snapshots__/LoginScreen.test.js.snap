// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <LoginScreen/> Debe hacer el Snapshot del componente 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <h1
          class="auth__title"
        >
          Login
        </h1>
        <form
          class="animate__animated animate__fadeIn animated__fater"
        >
          <input
            autocomplete="off"
            class="auth__input"
            name="email"
            placeholder="Email"
            type="text"
            value=""
          />
          <input
            autocomplete="off"
            class="auth__input"
            name="password"
            placeholder="Password"
            type="password"
            value=""
          />
          <button
            class="btn btn-primary btn-block"
            data-testid="btnLogin"
            type="submit"
          >
            Login
          </button>
          <hr />
          <div
            class="auth__social-netwoeks"
          >
            <p>
              Login with social network
            </p>
            <div
              class="google-btn"
              data-testid="withGoogle"
            >
              <div
                class="google-icon-wrapper"
              >
                <img
                  alt="google button"
                  class="google-icon"
                  src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                />
              </div>
              <p
                class="btn-text"
              >
                <b>
                  Sign in with google
                </b>
              </p>
            </div>
          </div>
          <a
            class="link"
            href="/auth/register"
          >
            Create new account
          </a>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <h1
        class="auth__title"
      >
        Login
      </h1>
      <form
        class="animate__animated animate__fadeIn animated__fater"
      >
        <input
          autocomplete="off"
          class="auth__input"
          name="email"
          placeholder="Email"
          type="text"
          value=""
        />
        <input
          autocomplete="off"
          class="auth__input"
          name="password"
          placeholder="Password"
          type="password"
          value=""
        />
        <button
          class="btn btn-primary btn-block"
          data-testid="btnLogin"
          type="submit"
        >
          Login
        </button>
        <hr />
        <div
          class="auth__social-netwoeks"
        >
          <p>
            Login with social network
          </p>
          <div
            class="google-btn"
            data-testid="withGoogle"
          >
            <div
              class="google-icon-wrapper"
            >
              <img
                alt="google button"
                class="google-icon"
                src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
              />
            </div>
            <p
              class="btn-text"
            >
              <b>
                Sign in with google
              </b>
            </p>
          </div>
        </div>
        <a
          class="link"
          href="/auth/register"
        >
          Create new account
        </a>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
